File: lua/trebuchet.lua
local aoc = require"adventofcode"

local function find_first_of(str, pats)
    local found_pos = #str + 1
    local found
    for _, pat in pairs(pats) do
        local pos, pose = string.find(str, pat)
        if pos and pos < found_pos then
            found = string.sub(str, pos, pose)
            found_pos = pos
        end
        -- small optimization if we have the first possible match
        if pos == 1 then
            break
        end
    end
    return found
end

local function find_last_of(str, pats)
    for pos = -1, -#str, -1 do
        for _, pat in pairs(pats) do
            local found = string.match(str, pat, pos)
            if found then return found end
        end
    end
    return nil
end

function trebuchet1 (line)
    line = string.lower(line)
    local first = string.match(line, '%d') or ''
    local last = string.match(string.reverse(line), '%d') or ''
    return tonumber(first .. last) or 0
end

local number_words = {
    'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'zero', '%d'
}
local function word_to_number (word)
    local num = tonumber(word)
    if num then return num end
    if word == 'zero' then return 0 end
    return aoc.index(number_words, word)
end

function trebuchet2 (line)
    line = string.lower(line)
    local first = find_first_of(line, number_words)
    local last = find_last_of(line, number_words)
    first = first and word_to_number(first) or ''
    last = last and word_to_number(last) or ''
    return tonumber(first .. last) or 0
end

-- main
local stage = ({['1']=trebuchet1, ['2']=trebuchet2})[arg[1]] or trebuchet1
local lines = {}
for line in io.lines() do
    lines[#lines+1] = stage(line)
end
print(aoc.reduce(lines, aoc.sum))

File: python/trebuchet.py
# Advent of Code 2023
# Day one task: Trepuchet?!

from adventofcode import AdventOfCode
import re

NUM = re.compile('\\d')
number_words = {
    'zero':0,
    'one':1,
    'two':2,
    'three':3,
    'four':4,
    'five':5,
    'six':6,
    'seven':7,
    'eight':8,
    'nine':9
}
NUM_WORDS = re.compile('|'.join(number_words)+'|\\d')
def word_to_int(word):
    return int(number_words.get(word,word))

def safe_search(match):
    return match.group(0) if match else ''

def find_all_overlapping(exp,line):
    start = 0
    while match := exp.search(line, start):
        yield match.group(0)
        start = match.start() + 1

class trebuchet1(AdventOfCode):

    def line(self, line):
        line = line.lower()
        return int(safe_search(NUM.search(line)) + safe_search(NUM.search(line[::-1])))

class trebuchet2(AdventOfCode):

    def line(self, line):
        # My first attempt here used `findall` but failed on 'eightwo'
        numbers = list(find_all_overlapping(NUM_WORDS, line.lower()))
        return word_to_int(numbers[0])*10 + word_to_int(numbers[-1]) if numbers else 0
    
def main():
    stage = trebuchet1
    import sys
    if len(sys.argv) > 1:
        if sys.argv[1] == '1':
            stage = trebuchet1
        elif sys.argv[1] == '2':
            stage = trebuchet2
        else:
            return
    print(stage(sys.stdin).reduce_lines(sum))

if __name__=='__main__':
    main()
